class-wxr-import-ui.php

```diff
@@ -0,0 +1,635 @@
+	 * @return bool|WP_Error True on success, error object otherwise.
+	 */
+	protected function handle_upload() {
+		$file = wp_import_handle_upload();
+
+		if ( isset( $file['error'] ) ) {
+			return new WP_Error( 'wxr_importer.upload.error', esc_html( $file['error'] ), $file );
+		} elseif ( ! file_exists( $file['file'] ) ) {
+			$message = sprintf(
+				__( 'The export file could not be found at <code>%s</code>. It is likely that this was caused by a permissions problem.', 'wordpress-importer' ),
+				esc_html( $file['file'] )
+			);
+			return new WP_Error( 'wxr_importer.upload.no_file', $message, $file );
+		}
+
+		$this->id = (int) $file['id'];
+		return true;
+	}
+
+	public function handle_async_upload() {
+		header( 'Content-Type: text/html; charset=' . get_option( 'blog_charset' ) );
+		send_nosniff_header();
+		nocache_headers();
+
+		check_ajax_referer( 'wxr-import-upload' );
+
+		/*
+		 * This function does not use wp_send_json_success() / wp_send_json_error()
+		 * as the html4 Plupload handler requires a text/html content-type for older IE.
+		 * See https://core.trac.wordpress.org/ticket/31037
+		 */
+
+		if ( ! current_user_can( 'upload_files' ) ) {
+			echo wp_json_encode( array(
+				'success' => false,
+				'data'    => array(
+					'message'  => __( 'You do not have permission to upload files.' ),
+					'filename' => $_FILES['import']['name'],
+				)
+			) );
+
+			exit;
+		}
+
+		$file = wp_import_handle_upload();
+		if ( is_wp_error( $file ) ) {
+			echo wp_json_encode( array(
+				'success' => false,
+				'data'    => array(
+					'message'  => $file->get_error_message(),
+					'filename' => $_FILES['import']['name'],
+				)
+			) );
+
+			wp_die();
+		}
+
+		if ( ! $attachment = wp_prepare_attachment_for_js( $file['id'] ) )
+			exit;
+
+		echo wp_json_encode( array(
+			'success' => true,
+			'data'    => $attachment,
+		) );
+
+		exit;
+	}
+
+	/**
+	 * Handle a WXR file selected from the media browser.
+	 *
+	 * @return bool|WP_Error True on success, error object otherwise.
+	 */
+	protected function handle_select( $id ) {
+		if ( ! is_numeric( $id ) || intval( $id ) < 1 ) {
+			return new WP_Error(
+				'wxr_importer.upload.invalid_id',
+				__( 'Invalid media item ID.', 'wordpress-importer' ),
+				compact( 'id' )
+			);
+		}
+
+		$id = (int) $id;
+
+		$attachment = get_post( $id );
+		if ( ! $attachment || $attachment->post_type !== 'attachment' ) {
+			return new WP_Error(
+				'wxr_importer.upload.invalid_id',
+				__( 'Invalid media item ID.', 'wordpress-importer' ),
+				compact( 'id', 'attachment' )
+			);
+		}
+
+		if ( ! current_user_can( 'read_post', $attachment->ID ) ) {
+			return new WP_Error(
+				'wxr_importer.upload.sorry_dave',
+				__( 'You cannot access the selected media item.', 'wordpress-importer' ),
+				compact( 'id', 'attachment' )
+			);
+		}
+
+		$this->id = $id;
+		return true;
+	}
+
+	protected function get_data_for_attachment( $id ) {
+		$existing = get_post_meta( $id, '_wxr_import_info' );
```

Comment:
Typically I do this because it means we can rely on `empty` meaning "no value set", since the return value is either `array()` or `array( $item )`. With single, there's no way to check not-set vs empty value.

In this case though, an empty value isn't usable, so can probably be simplified.